### PR TITLE
[Merged by Bors] - feat: port control.functor

### PR DESCRIPTION
**Pull request by j-loreaux**
mathlib3 SHA: 71ca477041bcd6d7c745fe555dc49735c12944b7

- [x] depends on: #588

Notes: 
1. There is a discrepancy between the Lean 3 and Lean 4 types of `has_seq.seq` and `Seq.seq`. I added an `ₓ` for alignment.
2. There were one or two places where the existing `simp` or `rw` calls didn't succeed, but I'm not entirely sure why.
3. In `Functor.Comp.functor_comp_id` and `Functor.Comp.functor_id_comp` I had to use `Id` instead of `id`, because only the former has the `Monad` instance.
4. I had to change the imports, but I think what is there now is minimal.

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#588
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#588~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Control/Functor.lean [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+  -- porting note: was `apply funext <;> intro <;> sorry --rw [comp_map]` but `rw` failed?
+#align functor.map_comp_map Functor.map_comp_map
+
+theorem Functor.ext {F} :
+    ∀ {F1 : Functor F} {F2 : Functor F} [@LawfulFunctor F F1] [@LawfulFunctor F F2]
+    (_H : ∀ (α β) (f : α → β) (x : F α), @Functor.map _ F1 _ _ f x = @Functor.map _ F2 _ _ f x),
+    F1 = F2
+  | ⟨m, mc⟩, ⟨m', mc'⟩, H1, H2, H => by
+    cases show @m = @m' by funext α β f x; apply H
+    -- porting note: Lean marks `H` as an unused variable?
```
Comment (semorrison) [inline Mathlib/Control/Functor.lean:55]: inaccessible, not unused. This is a hygiene issue. You can probably use `cases ... with ...` to make the name accessible.

File: Mathlib/Control/Functor.lean [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+variable [Functor F] [LawfulFunctor F]
+
+theorem Functor.map_id : (· <$> ·) id = (id : F α → F α) := by apply funext <;> apply id_map
+#align functor.map_id Functor.map_id
+
+theorem Functor.map_comp_map (f : α → β) (g : β → γ) :
+    ((· <$> ·) g ∘ (· <$> ·) f : F α → F γ) = (· <$> ·) (g ∘ f) :=
+  funext <| fun _ => (comp_map _ _ _).symm
+  -- porting note: was `apply funext <;> intro <;> sorry --rw [comp_map]` but `rw` failed?
```
Comment (semorrison) [inline Mathlib/Control/Functor.lean:46]:
```suggestion
  -- porting note: was `apply funext <;> intro <;> rw [comp_map]` but `rw` failed?
```

File: Mathlib/Control/Functor.lean (hunk 1 above)
Comment (j-loreaux) [reply Mathlib/Control/Functor.lean:55]: No, note that `H` is a hypothesis. It marked it as unused before I added the underscore.

Comment (semorrison) [reply Mathlib/Control/Functor.lean:55]: Oh, I see. That's actually perfectly reasonable. It's just looking at the forall in the return type, where indeed H is not used.

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
Other PRs have been getting through bors, so I propose we cross our fingers and hope this one will go through when we try again. Once the bors queue is clear, please feel free to try merging again!

bors d+

**Comment by j-loreaux** [comment]
one more try:

bors merge

**Pull request closed without merging**